### ISSUE TITLE
Metric parameter

### ISSUE THREAD

**Issue by heliosdrm** [JuliaDynamics/RecurrenceAnalysis.jl#10]
In #9 @Datseris suggested:

> [Optional/Suggestion] : Rework the handling of the metrics: instead of asking for a string ask for the `Metric` instance. E.g. `"euclidean" => Euclidean()`. This is important for education as it teaches the Julia way. But this is just my humble opinion.

If a funcion asks for a `Metric` instance, the user must be able to create it (e.g. calling `Euclidean()`). But this means that RecurrenceAnalysis.jl should export the `Metric` types, although they may have no other use than creating the parameters needed by the functions of this module. Is it not tidier to keep the `Metric` types unexported, and let the functions use them internally?

That is the reason why functions like `distancematrix`, etc. asked for strings, and a `Dict{String,Metric}` was used internally. If strings are bad style in Julia, perhaps symbols would be better (e.g. `:euclidean`, `:max`, etc.)?

In the notebook you also noted a risk of type instability with the current approach based on strings. Type instability pitfalls are still a bit obscure for me, so I'm probably missing something, but I don't understand why it may happen in this case: the variables that receive the strings are not used for the `Metric` objects in any part of the code. The string is just passed as a key of a dictionary of metrics, and the corresponding item of the dictionary is copied in a different variable.

**Comment by Datseris** [comment]
It makes no difference whether you use symbol or string, the type instability will stay the same. Notice that `typeof("euclidean") == typeof("chebyshev")` and the same goes if you use symbols. However,
`typeof(Euclidean()) != typeof(Chebyshev())`, because each distance is a different concrete type. 

This means that your dictionary is not concretely typed. All values are `Metric`, but `Metric` is abstract type. The compiler cannot deduce optimal compilation in this case. This is discussed (but only barely) in the documentation of Julia, performance tips, "avoid abstract containers".

Take the following example:
```
using Distances, StaticArrays, BenchmarkTools
const metrics = Dict("euclidean" =>Euclidean(), "cityblock" => Cityblock())

_get_metrics(s) = metrics[s]

function f1(a, b, distance::String)
    m = _get_metrics(distance)
    c = 0.0
    for (v1, v2) in zip(a, b)
        c += evaluate(m, v1, v2)
    end
    return c
end
function f2(a, b, m::Metric)
    c = 0.0
    for (v1, v2) in zip(a, b)
        c += evaluate(m, v1, v2)
    end
    return c
end
function f3(a, b, distance::String)
    m = _get_metrics(distance)
    return f2(a, b, m)
end

a = [rand(SVector{3}) for i in 1:1000]
b = [rand(SVector{3}) for i in 1:1000]

function bench(a, b, metric)
    M = _get_metrics(metric)
    @btime f1($a, $b, $metric)
    @btime f2($a, $b, $M)
    @btime f3($a, $b, $metric)
end

bench(a, b, "euclidean")
bench(a, b, "cityblock")
```
which outputs:
```
  58.298 μs (4000 allocations: 93.75 KiB)
  4.105 μs (0 allocations: 0 bytes)
  4.164 μs (1 allocation: 16 bytes)

  57.888 μs (4000 allocations: 93.75 KiB)
  1.478 μs (0 allocations: 0 bytes)
  1.560 μs (1 allocation: 16 bytes)
```

You can see such a massive difference in performance.

There are two ways forward: 1. you ask users to pass an instance of `Metric`, which is totally fine.

Or, 2, do what I do in the third function: create a function barrier. Have one function that creates the matrix and one more that actually does the compilation. Unfortunately, if you choose this way you will have to do it for **every function in your library that uses metrics**. (Otherwise you won't get good performance in all of them!)

There is another thing to consider here: code readability. My version needs 5 lines of code, which define `f2`. The "function barrier solution" requires a dictionary definition, and one more function definition. This may seem not too much in my trivial example, but over an entire library it adds up. 

**IMPORTANT** - It is _impossible_ to exactly reach the performance of `f2` using function barriers. As soon as your function speeds reaches the `μs` range, the dynamic dispatch (which takes on the order of a couple `μs`) will become apparent. It is already visible in my benchmarks, but for functions that are even faster this is even more severe. On the other hand, for functions that take on the order of ms the dynamic dispatch costs does not even matter and `f3` and `f2` have equal performance.

TL;DR: `f2` will always be the fastest possible version.

currently you use approach `f1`, which is really bad for performance. See e.g. in `distancematrix`:
```
function distancematrix(x::AbstractVecOrMat, metric::AbstractString="max")
    dist = getmetric(metric)   <---- this is type unstable
    pairwise(dist, x')  <---- compiler cannot optimize this operation
end
```

> But this means that RecurrenceAnalysis.jl should export the Metric types, although they may have no other use than creating the parameters needed by the functions of this module. Is it not tidier to keep the Metric types unexported, and let the functions use them internally?

I do not understand why this is a concern. First of all, _they do have a use for them_. Passing them to `recurrencematrix` or any other function! :) 
Secondly, because these metrics follow the camel-case convention (e.g. `Euclidean`), I do not believe that they pollute the name space.

---

Notice: I can understand that using strings instead of the type name `Euclidean()` could be better for a baby/beginner user. And in general I do like to bring simplicity to users. But it is up to you, the developer, to decide where to draw the line of how simple you want to make things. I personally think using the metric types instead of strings is a good choice because it leads to simpler code, but I could very easily live with using strings instead (provided that the `f3` approach is used).

**Comment by Datseris** [comment]
> Anyway, just for understanding, I fail to see the difference between the current approach:

They are two different functions. If they were the same they would have the same source code :)

The important difference is that the dynamic dispatch (i.e. at run time) happens only once inside `f3` in the line `return f2(a, b, m)` (hence only one single allocation in the benchmark). In `f1`, dynamic dispatch happens _every time_ on the line `c += evaluate(m, v1, v2)` as it is impossible for the compiler to deduce the type of `m` and thus create optimal code. Hence the 1000 allocations as the vectors happened to have 1000 elements.

Important knowledge: _all functions in Julia are compiled_. Functions that call other functions are _also compiled_. If you write bad code, it will compile badly, even if the individual functions are fast. Their fast version will not be compiled.

Function barriers are general Julia knowledge, not specific with our current case. Function barriers always help aleviate the performance deficit of type stability . Please read this page: https://docs.julialang.org/en/v1/manual/performance-tips/

especially [this](https://docs.julialang.org/en/v1/manual/performance-tips/#kernel-functions-1) and [this](https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-containers-with-abstract-type-parameters-1). Once you read this everything will be clear. At the moment I am just repeating this documentation, and it is'nt very efficient.

---

> The code of distancematrix is like f3 in your example, and the methods of Distances.pairwise are:

This shows confusion or lack of understanding of how Julia works, if I understood your comment correctly.

Please read the documentation on [Methods](https://docs.julialang.org/en/v1/manual/methods/). There you will find that these methods that cite above are only _declerations_. Julia creates a **specialized** function instance for each **concrete** subtype of `Metric`. If the `Metric` type is known at compile time, julia will produce superfast code. If the `Metric` type is known only at run-time, your code will be slow.

When you call a function, **all concrete types are fully known** as you can only instantiate concrete types. You can't just "call a function with abstract types". It is not possible nor does it make sense to say it.

---

> On the other hand, if calling functions with abstract types is bad for code specialization, using Union types might be an improvement:

Again same as above, this shows some confusion about the type system, because it is **fundamentally impossible** to call a function with abstract types. Please see the documentation on the type system of Julia: https://docs.julialang.org/en/v1/manual/types/

Unions are irrelevant with what we are discussing here and they should not be involved in any way.

---

> Not only for that, but also to make the interface of the functions in RecurrenceAnalysis independent of Distances.jl. So we may prevent breaking changes if in the future we want to change the way of computing distances (e.g. using another package), or if Distances.jl makes some breaking change itself.

Okay. As I said, moving forward with the approach of `f3` is okay by me. I can live with it (`f2` still better tho :D ). Keeping the `f1` approach is impossible though, if you actually care about performance.

**Comment by heliosdrm** [comment]
Thanks a lot! This is very instructive and helpful.

If I may abuse of your patience, I'd ask again why `distancematrix` does not follow the model of `f3` (or the example in the documentation about function barriers). It just does:

1. `dist = getmetric(metric)`. Yes, this is type unstable, that I understand clearly now.
2. `pairwise(dist, x')`. This is the function barrier, isn't it?

If there is any kind of loop, that's inside `pairwise`, which does know the type of `dist`, right?

I have tried a version of `distancematrix` that gets metric directly as a `Metric` instance (no dictionaries, etc.), and the performance loss is negligible...

**Comment by Datseris** [comment]
Ops , sorry, you are correct. I made a mistake, what I said was wrong.
```
function distancematrix(x::AbstractVecOrMat, metric::AbstractString="max")
    dist = getmetric(metric)   <---- this is type unstable
    pairwise(dist, x')  <---- compiler CAN optimize this operation. it is the function barrier
end
```
I was being absent minded, thinking of the version of `distancematrix` that _I wrote_ in the jupyter notebook. There we would have a problem.

And I think we want to have both versions to support `Dataset`, so least the version of my notebook has to be separated into function barrier.
